common/src/lru.rs

```diff
@@ -1,925 +1,286 @@
-    }
-}
-
-impl<K: PartialEq> PartialEq for KeyRef<K> {
-    fn eq(&self, other: &KeyRef<K>) -> bool {
-        unsafe { (*self.k).eq(&*other.k) }
-    }
-}
+use crate::HashMap;
+use alloc::{collections::VecDeque, sync::Arc, vec::Vec};
+use core::hash::Hash;
 
-impl<K: Eq> Eq for KeyRef<K> {}
+/// LRU Cache.
+pub struct LruCache<K, V> {
+    /// Entries currently in cache.
+    entries: Vec<Option<(Arc<K>, V)>>,
 
-#[doc(hidden)]
-pub auto trait NotKeyRef {}
+    /// A map of keys -> their index in the `entries` vector, used to speed up lookups.
+    key_to_entry_index: HashMap<Arc<K>, usize>,
 
-impl<K> !NotKeyRef for KeyRef<K> {}
+    /// Indexes of used entries inside the `entries` array. Sorted from newest to oldest.
+    used_indexes: VecDeque<usize>,
 
-impl<K, D> Borrow<D> for KeyRef<K>
-where
-    K: Borrow<D>,
-    D: NotKeyRef + ?Sized,
-{
-    fn borrow(&self) -> &D {
-        unsafe { &*self.k }.borrow()
-    }
+    /// Indexes of free entries in the `entries` array.
+    free_indexes: VecDeque<usize>,
 }
 
-// Struct used to hold a key value pair. Also contains references to previous and next entries
-// so we can maintain the entries in a linked list ordered by their use.
-struct LruEntry<K, V> {
-    key: K,
-    val: V,
-    prev: *mut LruEntry<K, V>,
-    next: *mut LruEntry<K, V>,
-}
+impl<K: PartialEq + Eq + Hash, V> LruCache<K, V> {
+    /// Create a new LRU cache instance.
+    pub fn new(capacity: usize) -> Self {
+        let mut entries = Vec::with_capacity(capacity);
+        let mut free_indexes = VecDeque::with_capacity(capacity);
+        for i in 0..capacity {
+            entries.push(None);
+            free_indexes.push_back(i);
+        }
 
-impl<K, V> LruEntry<K, V> {
-    fn new(key: K, val: V) -> Self {
-        LruEntry {
-            key,
-            val,
-            prev: ptr::null_mut(),
-            next: ptr::null_mut(),
+        Self {
+            entries,
+            key_to_entry_index: HashMap::default(),
+            used_indexes: VecDeque::with_capacity(capacity),
+            free_indexes,
         }
     }
-}
-
-pub type DefaultHasher = crate::hasher_builder::HasherBuilder;
-
-/// An LRU Cache
-pub struct LruCache<K, V, S = DefaultHasher> {
-    map: HashMap<KeyRef<K>, Box<LruEntry<K, V>>, S>,
-    cap: usize,
-
-    // head and tail are sigil nodes to faciliate inserting entries
-    head: *mut LruEntry<K, V>,
-    tail: *mut LruEntry<K, V>,
-}
 
-impl<K: Hash + Eq, V> LruCache<K, V> {
-    /// Creates a new LRU Cache that holds at most `cap` items.
-    ///
-    /// # Example
-    ///
-    /// ```
-    /// use mc_common::LruCache;
-    /// let mut cache: LruCache<isize, &str> = LruCache::new(10);
-    /// ```
-    pub fn new(cap: usize) -> LruCache<K, V> {
-        LruCache::construct(
-            cap,
-            HashMap::with_capacity_and_hasher(cap, DefaultHasher::default()),
-        )
+    /// Returns the number of elements in the cache.
+    pub fn len(&self) -> usize {
```

Comment:
Should this maybe return a `Result<usize, Error>` rather than asserting the lengths of internal members match?

Reply:
I don't think so. This asset failing means there is a bug in the code - it should never happen.